nck/clients/sa360_client.py

```diff
@@ -0,0 +1,151 @@
+            access_token=access_token,
+            client_id=client_id,
+            client_secret=client_secret,
+            refresh_token=refresh_token,
+            token_expiry=None,
+            token_uri=GOOGLE_TOKEN_URI,
+            user_agent=None,
+        )
+        http = self._credentials.authorize(httplib2.Http())
+        self._credentials.refresh(http)
+        self.auth = (
+            f"{self._credentials.token_response['token_type']} {self._credentials.token_response['access_token']}"
+        )
+        self._service = discovery.build(self.API_NAME, self.API_VERSION, http=http, cache_discovery=False)
+
+    def get_all_advertisers_of_agency(self, agency_id):
+        body = {
+            "reportScope": {"agencyId": agency_id},
+            "reportType": "advertiser",
+            "columns": [{"columnName": "advertiserId"}],
+            "statisticsCurrency": "usd",
+        }
+        report = self._service.reports().generate(body=body).execute()
+        advertiser_ids = [row["advertiserId"] for row in report["rows"]]
+        return advertiser_ids
+
+    @staticmethod
+    def generate_report_body(
+        agency_id, advertiser_id, report_type, columns, start_date, end_date, custom_metrics, custom_dimensions
+    ):
+        all_columns = SA360Client.generate_columns(columns, custom_metrics, custom_dimensions)
+        body = {
+            "reportScope": {"agencyId": agency_id, "advertiserId": advertiser_id},
+            "reportType": report_type,
+            "columns": all_columns,
+            "timeRange": SA360Client.get_date_range(start_date, end_date),
+            "downloadFormat": "csv",
+            "maxRowsPerFile": 4000000,
+            "statisticsCurrency": "usd",
+        }
+        logger.info("Report Body Generated")
+
+        return body
+
+    def request_report_id(self, body):
+        report = self._service.reports().request(body=body).execute()
+        logger.info("Report requested!")
+        return report["id"]
+
+    @retry(wait=wait_exponential(multiplier=60, min=60, max=600), stop=stop_after_delay(3600))
+    def assert_report_file_ready(self, report_id):
+        """Poll the API with the reportId until the report is ready, up to 100 times.
+
+               Args:
+                 report_id: The ID SA360 has assigned to a report.
+               """
+        request = self._service.reports().get(reportId=report_id)
+        report_data = request.execute()
+        if report_data["isReportReady"]:
+            logger.info("The report is ready.")
+
+            # For large reports, SA360 automatically fragments the report into multiple
+            # files. The 'files' property in the JSON object that SA360 returns contains
+            # the list of URLs for file fragment. To download a report, SA360 needs to
+            # know the report ID and the index of a file fragment.
+            return report_data
+        else:
+            logger.info("Report is not ready.")
```

Comment:
Could we add information on how many retry were already done?
Ex: "Report is not ready - tentative 7 - retry in 30 seconds" (if it's possible with the @retry decorator)

Reply:
seems like it's more complicated with the wait_exponential, so not necessary

https://stackoverflow.com/questions/55273453/is-there-a-way-to-log-number-of-retries-made

Reply:
yes, with an exponential delay it seems to be more difficult, it's not mandatory for a v0 I think